src/main/scala/io/iohk/ethereum/db/storage/ReferenceCountNodeStorage.scala

```diff
@@ -0,0 +1,132 @@
+package io.iohk.ethereum.db.storage
+
+import akka.util.ByteString
+import io.iohk.ethereum.db.storage.NodeStorage.{NodeEncoded, NodeHash}
+import io.iohk.ethereum.mpt.NodesKeyValueStorage
+import io.iohk.ethereum.rlp.RLPImplicitConversions.{toRlpList, _}
+import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{encode => rlpEncode, _}
+
+/**
+  * This class is used to store Nodes (defined in mpt/Node.scala), by using:
+  * Key: hash of the RLP encoded node
+  * Value: the RLP encoded node
+  */
+class ReferenceCountNodeStorage(nodeStorage: NodeStorage, blockNumber: Option[BigInt] = None) extends NodesKeyValueStorage {
+
+  import ReferenceCountNodeStorage._
+
+  override def get(key: ByteString): Option[NodeEncoded] = nodeStorage.get(key).map(_.toStoredNode).map(_.nodeEncoded.toArray)
+
+  override def update(toRemove: Seq[NodeHash], toUpsert: Seq[(NodeHash, NodeEncoded)]): NodesKeyValueStorage = {
+
+    require(blockNumber.isDefined)
+
+    val bn = blockNumber.get
```

Comment:
what to you think about making it non-optional in the constructor? is there a reason not to do so?